### ISSUE TITLE
False positive E3004 circular dependency for SAM templates

### ISSUE THREAD

**Issue by michael-k** [aws-cloudformation/cfn-lint#2501]
### CloudFormation Lint Version

cfn-lint 0.72.0

### What operating system are you using?

Linux

### Describe the bug

When `AWS::Serverless::Function` references an alarm in its `DeploymentPreference.Alarms` and the `AWS::CloudWatch::Alarm` references the function (eg. in Dimensions), a `E3004 Circular dependency` error is reported by cfn-lint.  Afaik this is a valid use case.

This was introduced in cfn-lint 0.72.0 (0.71.1 did not report E3004), most likely as part of PR aws-cloudformation/cfn-lint#2452.

```
E3004 Circular Dependencies for resource TestFunction. Circular dependency with [LatestVersionErrorMetricGreaterThanZeroAlarm]
template.yaml:5:3

E3004 Circular Dependencies for resource LatestVersionErrorMetricGreaterThanZeroAlarm. Circular dependency with [TestFunction]
template.yaml:28:3
```

(The template is based on https://gist.github.com/alexcasalboni/242071568a6adde805027345cfbb6d99.  But we first noticed this in one of our production functions when upgrading cfn-lint to 0.72.0.)

### Expected behavior

No E3004 is reported.

### Reproduction template

```yaml
AWSTemplateFormatVersion: '2010-09-09'
Transform: 'AWS::Serverless-2016-10-31'
Resources:

  TestFunction:
    Type: AWS::Serverless::Function
    Properties:
      Handler: index.test
      Runtime: nodejs16.x
      MemorySize: 128
      Timeout: 30
      Tracing: Active
      CodeUri: ./
      Events:
        GetResource:
          Type: Api
          Properties:
            Path: /test
            Method: get
            RestApiId: !Ref 'Api'
      
      AutoPublishAlias: live
      DeploymentPreference:
        Type: Linear10PercentEvery1Minute
        Alarms:
          - !Ref LatestVersionErrorMetricGreaterThanZeroAlarm

  LatestVersionErrorMetricGreaterThanZeroAlarm:
    Type: AWS::CloudWatch::Alarm
    Properties:
      AlarmName: alarm-sam-demo-latestversion
      AlarmDescription: "alarm to check for errors in the function"
      ComparisonOperator: GreaterThanOrEqualToThreshold
      TreatMissingData: missing
      Dimensions:
        - Name: FunctionName
          Value: !Ref TestFunction
        - Name: Resource
          Value: !Join [":", [!Ref TestFunction, !Select ["7", !Split [":", !Ref TestFunction.Version]]]]
      EvaluationPeriods: 1
      MetricName: Errors
      Namespace: AWS/Lambda
      Period: 60
      Statistic: Sum
      Threshold: 1

  Api:
    Type: AWS::Serverless::Api
    Properties:
      StageName: foo
      TracingEnabled: true

  LogGroup:
    Type: AWS::Logs::LogGroup
    Properties:
      LogGroupName: !Sub
        - /aws/lambda/${FunctionName}
        - FunctionName: !Ref 'TestFunction'
      RetentionInDays: 1
    DeletionPolicy: Delete
    UpdateReplacePolicy: Delete
```

**Comment by amandalishus** [comment]
I was about to create an issue for a "bug" I found, based on advice from @cmgorton (I assume you're ChristinaG?) on Discord, but it seems almost certain that it is connected to this issue; if I had to create a title for it, I would have said something like "False Positive `E3004 Circular Dependency` for `AWS::Serverless::Api.Variables`", which appears suspiciously similar to this one. I say "bug" somewhat hesitantly, because it's not _strictly_ a problem with CloudFormation; rather, it's a bug that occurs when cfn-lint is parsing CloudFormation that includes a (thrice-cursed) Serverless Application Model template. Personally, I would prefer to remove SAM resources from our CF completely, but I don't have that option right now... and it seems like I'm not the only one who is expecting cfn-lint to handle SAM templates as well. Sorry 'bout that...

Anyway, the problem occurs when defining an `AWS::Serverless::Api` resource for an API with `EndpointConfiguration: Type: "PRIVATE"`. We do not have a custom DNS name (since the API is private), so we need to build the base URI and pass it to the OpenAPI 3.0 definition. The first part of the URI is the ID of the `AWS::ApiGateway::RestApi` resource which is created by the transformed SAM template; unfortunately, the SAM translation renders the name of the `AWS::ApiGateway::RestApi` as the same name used for the `AWS::Serverless::Api`, which causes a E3004 Circular Dependency.

Even knowing this works, the error appears perfectly logical; the ID of the resource is being referenced inside the definition of the resource. It's only due to my familiarity with the SAM translation (due to many, many issues like this, leading to my disdain for the model) that I was able to parse out why it works... the `Variables` section of the `AWS::Serverless::Api` resource unprocessed as the `Variables` section of the `AWS::ApiGateway::Stage` resource, which is already necessarily dependent on the `AWS::ApiGateway::RestApi` resource.

Since I was planning on reporting this as an issue myself (rather than a me-too comment), I put together a template which reproduces the issue, and I figured that still might help so I'm attached it 
[here](https://github.com/aws-cloudformation/cfn-lint/files/10094600/example.yaml.txt). Honestly though, now that I realize @michael-k beat me to the report, I suspect this issue lurks in several other disused corners of the SAM transformation. I know I've found at least two errors in the `aws://sam.schema.json` file which other tools (notably the RedHat YAML Extension for VSCode) use to validate SAM resources, and while those would be _easy_ to fix I have no idea where to report them. OTOH, I thought cfn-lint transformed the SAM resources into CF prior to validation, which leaves me puzzled about how this happens.

Anyway, if I was a little more brave, if I was any less puzzled, or if this bug had been sitting around for months already, I might jump into the code and try to find it; I'd love to get more involved with a project like this, but I have a feeling somebody involved in the last Pull Request is smacking themselves in the forehead right now so I'll leave this one to the experts. If anyone can point me to a root location for the omni-present `sam.schema.json` file, however, I could fix a couple bugs there in less than 10 minutes 😉 

Good luck! And thank you for all the work you put into these tools so that fools like me can find a way to break them 😆
